File: convert_json_to_excel.py
#!/usr/bin/env python3
"""
Convert IM8 JSON assessment files to Excel format
Usage: python convert_json_to_excel.py
"""

import json
import openpyxl
from openpyxl.styles import Font, PatternFill, Alignment, Border, Side
from openpyxl.utils import get_column_letter
from datetime import datetime
import os

def create_excel_from_json(json_file_path, output_excel_path):
    """Convert IM8 JSON to formatted Excel workbook"""
    
    print(f"📖 Reading JSON file: {json_file_path}")
    with open(json_file_path, 'r', encoding='utf-8') as f:
        data = json.load(f)
    
    # Create workbook
    wb = openpyxl.Workbook()
    wb.remove(wb.active)  # Remove default sheet
    
    # Define styles
    header_fill = PatternFill(start_color="366092", end_color="366092", fill_type="solid")
    header_font = Font(bold=True, color="FFFFFF", size=11)
    domain_fill = PatternFill(start_color="4472C4", end_color="4472C4", fill_type="solid")
    domain_font = Font(bold=True, color="FFFFFF", size=12)
    cell_border = Border(
        left=Side(style='thin'),
        right=Side(style='thin'),
        top=Side(style='thin'),
        bottom=Side(style='thin')
    )
    
    # Create Cover Page
    print("📄 Creating Cover Page...")
    cover_sheet = wb.create_sheet("Cover Page", 0)
    cover_data = data.get('cover_page', {})
    
    cover_sheet['A1'] = cover_data.get('title', 'IM8 Assessment')
    cover_sheet['A1'].font = Font(bold=True, size=16)
    
    cover_info = [
        ['Agency:', cover_data.get('agency_name', '')],
        ['Assessment Period:', cover_data.get('assessment_period', '')],
        ['Prepared By:', cover_data.get('prepared_by', '')],
        ['Preparation Date:', cover_data.get('preparation_date', '')],
        ['Reviewed By:', cover_data.get('reviewed_by', '')],
        ['Review Date:', cover_data.get('review_date', '')],
        ['Approval Status:', cover_data.get('approval_status', '')]
    ]
    
    row = 3
    for label, value in cover_info:
        cover_sheet[f'A{row}'] = label
        cover_sheet[f'A{row}'].font = Font(bold=True)
        cover_sheet[f'B{row}'] = value
        row += 1
    
    # Add instructions if present
    if 'instructions' in data:
        instructions = data['instructions']
        cover_sheet[f'A{row+1}'] = 'Instructions:'
        cover_sheet[f'A{row+1}'].font = Font(bold=True, size=12)
        row += 2
        
        cover_sheet[f'A{row}'] = instructions.get('overview', '')
        row += 2
        
        if 'completion_steps' in instructions:
            cover_sheet[f'A{row}'] = 'Completion Steps:'
            cover_sheet[f'A{row}'].font = Font(bold=True)
            row += 1
            for step in instructions['completion_steps']:
                cover_sheet[f'A{row}'] = step
                row += 1
    
    # Adjust column widths
    cover_sheet.column_dimensions['A'].width = 20
    cover_sheet.column_dimensions['B'].width = 50
    
    # Create domain sheets
    domains = data.get('domains', [])
    print(f"📊 Creating {len(domains)} domain sheets...")
    
    for domain in domains:
        domain_id = domain.get('domain_id', 'Unknown')
        domain_name = domain.get('domain_name', 'Unknown')
        sheet_name = f"{domain_id}"[:31]  # Excel sheet name limit
        
        print(f"  → {sheet_name}: {domain_name}")
        ws = wb.create_sheet(sheet_name)
        
        # Domain header
        ws['A1'] = f"{domain_id}: {domain_name}"
        ws['A1'].font = domain_font
        ws['A1'].fill = domain_fill
        ws['A1'].alignment = Alignment(horizontal='left', vertical='center')
        ws.merge_cells('A1:H1')
        
        ws['A2'] = domain.get('description', '')
        ws['A2'].alignment = Alignment(wrap_text=True)
        ws.merge_cells('A2:H2')
        ws.row_dimensions[2].height = 30
        
        # Control table headers
        headers = ['Control ID', 'Control Name', 'Description', 'Implementation Status', 
                   'Evidence Reference', 'Evidence Files', 'Notes', 'Last Review']
        
        for col_num, header in enumerate(headers, 1):
            cell = ws.cell(row=4, column=col_num)
            cell.value = header
            cell.font = header_font
            cell.fill = header_fill
            cell.alignment = Alignment(horizontal='center', vertical='center', wrap_text=True)
            cell.border = cell_border
        
        # Add controls
        controls = domain.get('controls', [])
        row = 5
        
        for control in controls:
            ws.cell(row=row, column=1).value = control.get('control_id', '')
            ws.cell(row=row, column=2).value = control.get('control_name', '')
            ws.cell(row=row, column=3).value = control.get('description', '')
            ws.cell(row=row, column=4).value = control.get('implementation_status', '')
            ws.cell(row=row, column=5).value = control.get('evidence_reference', '')
            
            # Evidence files
            evidence_files = control.get('evidence_files', [])
            if evidence_files:
                files_text = '\n'.join([f"• {ef.get('filename', ef) if isinstance(ef, dict) else ef}" 
                                       for ef in evidence_files])
                ws.cell(row=row, column=6).value = files_text
            else:
                ws.cell(row=row, column=6).value = ''
            
            ws.cell(row=row, column=7).value = control.get('notes', '')
            ws.cell(row=row, column=8).value = control.get('last_review_date', '')
            
            # Apply borders and alignment
            for col in range(1, 9):
                cell = ws.cell(row=row, column=col)
                cell.border = cell_border
                cell.alignment = Alignment(vertical='top', wrap_text=True)
            
            ws.row_dimensions[row].height = 60
            row += 1
        
        # Set column widths
        ws.column_dimensions['A'].width = 12
        ws.column_dimensions['B'].width = 30
        ws.column_dimensions['C'].width = 40
        ws.column_dimensions['D'].width = 18
        ws.column_dimensions['E'].width = 20
        ws.column_dimensions['F'].width = 25
        ws.column_dimensions['G'].width = 40
        ws.column_dimensions['H'].width = 12
    
    # Create Summary Dashboard
    print("📈 Creating Summary Dashboard...")
    summary_sheet = wb.create_sheet("Summary Dashboard")
    summary_data = data.get('summary_dashboard', {})
    
    summary_sheet['A1'] = 'IM8 Assessment Summary Dashboard'
    summary_sheet['A1'].font = Font(bold=True, size=14)
    summary_sheet.merge_cells('A1:B1')
    
    summary_items = [
        ['Total Controls:', summary_data.get('total_controls', 0)],
        ['Completion Percentage:', f"{summary_data.get('completion_percentage', 0)}%"],
        ['Domains Completed:', summary_data.get('domains_completed', 0)],
        ['Evidence Items Attached:', summary_data.get('evidence_items_attached', 0)],
        ['Validation Status:', summary_data.get('validation_status', '')],
        ['Last Updated:', summary_data.get('last_updated', '')],
        ['Submitted By:', summary_data.get('submitted_by', '')],
        ['Submission Date:', summary_data.get('submission_date', '')]
    ]
    
    row = 3
    for label, value in summary_items:
        summary_sheet[f'A{row}'] = label
        summary_sheet[f'A{row}'].font = Font(bold=True)
        summary_sheet[f'B{row}'] = value
        row += 1
    
    # Controls by status
    if 'controls_by_status' in summary_data:
        row += 1
        summary_sheet[f'A{row}'] = 'Controls by Status:'
        summary_sheet[f'A{row}'].font = Font(bold=True, size=12)
        row += 1
        
        for status, count in summary_data['controls_by_status'].items():
            summary_sheet[f'A{row}'] = status.replace('_', ' ').title()
            summary_sheet[f'B{row}'] = count
            row += 1
    
    summary_sheet.column_dimensions['A'].width = 25
    summary_sheet.column_dimensions['B'].width = 30
    
    # Save workbook
    print(f"💾 Saving Excel file: {output_excel_path}")
    wb.save(output_excel_path)
    print(f"✅ Excel file created successfully!")
    return output_excel_path


def main():
    """Main conversion function"""
    print("=" * 60)
    print("IM8 JSON to Excel Converter")
    print("=" * 60)
    print()
    
    # Check if openpyxl is installed
    try:
        import openpyxl
    except ImportError:
        print("❌ Error: openpyxl library not found!")
        print("📦 Installing openpyxl...")
        import subprocess
        subprocess.check_call(['pip', 'install', 'openpyxl'])
        print("✅ openpyxl installed successfully!")
        print()
    
    # File paths
    template_json = 'storage/IM8_Assessment_Template.json'
    completed_json = 'storage/IM8_Security_Awareness_Training_Completed.json'
    
    template_excel = 'storage/IM8_Assessment_Template.xlsx'
    completed_excel = 'storage/IM8_Security_Awareness_Training_Completed.xlsx'
    
    # Convert template
    if os.path.exists(template_json):
        print("🔄 Converting Template JSON to Excel...")
        create_excel_from_json(template_json, template_excel)
        print()
    else:
        print(f"⚠️  Template JSON not found: {template_json}")
        print()
    
    # Convert completed assessment
    if os.path.exists(completed_json):
        print("🔄 Converting Completed Assessment JSON to Excel...")
        create_excel_from_json(completed_json, completed_excel)
        print()
    else:
        print(f"⚠️  Completed JSON not found: {completed_json}")
        print()
    
    print("=" * 60)
    print("✅ Conversion Complete!")
    print("=" * 60)
    print()
    print("📁 Generated Files:")
    if os.path.exists(template_excel):
        print(f"  1. {template_excel}")
    if os.path.exists(completed_excel):
        print(f"  2. {completed_excel}")
    print()
    print("📝 Next Steps:")
    print("  1. Open the Excel files to review the structure")
    print("  2. Template file can be shared with Analysts")
    print("  3. Completed file shows example submission with data")
    print()


if __name__ == '__main__':
    main()

File: convert_json_to_excel.js
const fs = require('fs');
const ExcelJS = require('exceljs');

async function createExcelFromJson(jsonFilePath, outputExcelPath) {
    console.log(`📖 Reading JSON file: ${jsonFilePath}`);
    const data = JSON.parse(fs.readFileSync(jsonFilePath, 'utf8'));
    
    const workbook = new ExcelJS.Workbook();
    
    // Create Cover Page
    console.log('📄 Creating Cover Page...');
    const coverSheet = workbook.addWorksheet('Cover Page');
    const coverData = data.cover_page || {};
    
    coverSheet.getCell('A1').value = coverData.title || 'IM8 Assessment';
    coverSheet.getCell('A1').font = { bold: true, size: 16 };
    
    const coverInfo = [
        ['Agency:', coverData.agency_name || ''],
        ['Assessment Period:', coverData.assessment_period || ''],
        ['Prepared By:', coverData.prepared_by || ''],
        ['Preparation Date:', coverData.preparation_date || ''],
        ['Reviewed By:', coverData.reviewed_by || ''],
        ['Review Date:', coverData.review_date || ''],
        ['Approval Status:', coverData.approval_status || '']
    ];
    
    let row = 3;
    coverInfo.forEach(([label, value]) => {
        coverSheet.getCell(`A${row}`).value = label;
        coverSheet.getCell(`A${row}`).font = { bold: true };
        coverSheet.getCell(`B${row}`).value = value;
        row++;
    });
    
    // Add instructions
    if (data.instructions) {
        const instructions = data.instructions;
        coverSheet.getCell(`A${row + 1}`).value = 'Instructions:';
        coverSheet.getCell(`A${row + 1}`).font = { bold: true, size: 12 };
        row += 2;
        
        coverSheet.getCell(`A${row}`).value = instructions.overview || '';
        coverSheet.getCell(`A${row}`).alignment = { wrapText: true };
        row += 2;
        
        if (instructions.completion_steps) {
            coverSheet.getCell(`A${row}`).value = 'Completion Steps:';
            coverSheet.getCell(`A${row}`).font = { bold: true };
            row++;
            instructions.completion_steps.forEach(step => {
                coverSheet.getCell(`A${row}`).value = step;
                row++;
            });
        }
    }
    
    coverSheet.getColumn('A').width = 20;
    coverSheet.getColumn('B').width = 50;
    
    // Create domain sheets
    const domains = data.domains || [];
    console.log(`📊 Creating ${domains.length} domain sheets...`);
    
    for (const domain of domains) {
        const domainId = domain.domain_id || 'Unknown';
        const domainName = domain.domain_name || 'Unknown';
        const sheetName = domainId.substring(0, 31);
        
        console.log(`  → ${sheetName}: ${domainName}`);
        const ws = workbook.addWorksheet(sheetName);
        
        // Domain header
        ws.getCell('A1').value = `${domainId}: ${domainName}`;
        ws.getCell('A1').font = { bold: true, color: { argb: 'FFFFFFFF' }, size: 12 };
        ws.getCell('A1').fill = {
            type: 'pattern',
            pattern: 'solid',
            fgColor: { argb: 'FF4472C4' }
        };
        ws.mergeCells('A1:H1');
        
        ws.getCell('A2').value = domain.description || '';
        ws.getCell('A2').alignment = { wrapText: true };
        ws.mergeCells('A2:H2');
        ws.getRow(2).height = 30;
        
        // Headers
        const headers = ['Control ID', 'Control Name', 'Description', 'Implementation Status', 
                        'Evidence Reference', 'Evidence Files', 'Notes', 'Last Review'];
        
        headers.forEach((header, idx) => {
            const cell = ws.getCell(4, idx + 1);
            cell.value = header;
            cell.font = { bold: true, color: { argb: 'FFFFFFFF' } };
            cell.fill = {
                type: 'pattern',
                pattern: 'solid',
                fgColor: { argb: 'FF366092' }
            };
            cell.alignment = { horizontal: 'center', vertical: 'middle', wrapText: true };
            cell.border = {
                top: { style: 'thin' },
                left: { style: 'thin' },
                bottom: { style: 'thin' },
                right: { style: 'thin' }
            };
        });
        
        // Add controls
        const controls = domain.controls || [];
        let rowNum = 5;
        
        for (const control of controls) {
            ws.getCell(rowNum, 1).value = control.control_id || '';
            ws.getCell(rowNum, 2).value = control.control_name || '';
            ws.getCell(rowNum, 3).value = control.description || '';
            ws.getCell(rowNum, 4).value = control.implementation_status || '';
            ws.getCell(rowNum, 5).value = control.evidence_reference || '';
            
            // Evidence files
            const evidenceFiles = control.evidence_files || [];
            if (evidenceFiles.length > 0) {
                const filesText = evidenceFiles.map(ef => 
                    `• ${typeof ef === 'object' ? ef.filename : ef}`
                ).join('\n');
                ws.getCell(rowNum, 6).value = filesText;
            }
            
            ws.getCell(rowNum, 7).value = control.notes || '';
            ws.getCell(rowNum, 8).value = control.last_review_date || '';
            
            // Apply borders and alignment
            for (let col = 1; col <= 8; col++) {
                const cell = ws.getCell(rowNum, col);
                cell.border = {
                    top: { style: 'thin' },
                    left: { style: 'thin' },
                    bottom: { style: 'thin' },
                    right: { style: 'thin' }
                };
                cell.alignment = { vertical: 'top', wrapText: true };
            }
            
            ws.getRow(rowNum).height = 60;
            rowNum++;
        }
        
        // Set column widths
        ws.getColumn(1).width = 12;
        ws.getColumn(2).width = 30;
        ws.getColumn(3).width = 40;
        ws.getColumn(4).width = 18;
        ws.getColumn(5).width = 20;
        ws.getColumn(6).width = 25;
        ws.getColumn(7).width = 40;
        ws.getColumn(8).width = 12;
    }
    
    // Create Summary Dashboard
    console.log('📈 Creating Summary Dashboard...');
    const summarySheet = workbook.addWorksheet('Summary Dashboard');
    const summaryData = data.summary_dashboard || {};
    
    summarySheet.getCell('A1').value = 'IM8 Assessment Summary Dashboard';
    summarySheet.getCell('A1').font = { bold: true, size: 14 };
    summarySheet.mergeCells('A1:B1');
    
    const summaryItems = [
        ['Total Controls:', summaryData.total_controls || 0],
        ['Completion Percentage:', `${summaryData.completion_percentage || 0}%`],
        ['Domains Completed:', summaryData.domains_completed || 0],
        ['Evidence Items Attached:', summaryData.evidence_items_attached || 0],
        ['Validation Status:', summaryData.validation_status || ''],
        ['Last Updated:', summaryData.last_updated || ''],
        ['Submitted By:', summaryData.submitted_by || ''],
        ['Submission Date:', summaryData.submission_date || '']
    ];
    
    row = 3;
    summaryItems.forEach(([label, value]) => {
        summarySheet.getCell(`A${row}`).value = label;
        summarySheet.getCell(`A${row}`).font = { bold: true };
        summarySheet.getCell(`B${row}`).value = value;
        row++;
    });
    
    // Controls by status
    if (summaryData.controls_by_status) {
        row++;
        summarySheet.getCell(`A${row}`).value = 'Controls by Status:';
        summarySheet.getCell(`A${row}`).font = { bold: true, size: 12 };
        row++;
        
        Object.entries(summaryData.controls_by_status).forEach(([status, count]) => {
            summarySheet.getCell(`A${row}`).value = status.replace(/_/g, ' ').replace(/\b\w/g, l => l.toUpperCase());
            summarySheet.getCell(`B${row}`).value = count;
            row++;
        });
    }
    
    summarySheet.getColumn('A').width = 25;
    summarySheet.getColumn('B').width = 30;
    
    // Save workbook
    console.log(`💾 Saving Excel file: ${outputExcelPath}`);
    await workbook.xlsx.writeFile(outputExcelPath);
    console.log(`✅ Excel file created successfully!`);
}

async function main() {
    console.log('='.repeat(60));
    console.log('IM8 JSON to Excel Converter (Node.js)');
    console.log('='.repeat(60));
    console.log();
    
    const templateJson = 'storage/IM8_Assessment_Template.json';
    const completedJson = 'storage/IM8_Security_Awareness_Training_Completed.json';
    
    const templateExcel = 'storage/IM8_Assessment_Template.xlsx';
    const completedExcel = 'storage/IM8_Security_Awareness_Training_Completed.xlsx';
    
    try {
        // Convert template
        if (fs.existsSync(templateJson)) {
            console.log('🔄 Converting Template JSON to Excel...');
            await createExcelFromJson(templateJson, templateExcel);
            console.log();
        } else {
            console.log(`⚠️  Template JSON not found: ${templateJson}`);
            console.log();
        }
        
        // Convert completed assessment
        if (fs.existsSync(completedJson)) {
            console.log('🔄 Converting Completed Assessment JSON to Excel...');
            await createExcelFromJson(completedJson, completedExcel);
            console.log();
        } else {
            console.log(`⚠️  Completed JSON not found: ${completedJson}`);
            console.log();
        }
        
        console.log('='.repeat(60));
        console.log('✅ Conversion Complete!');
        console.log('='.repeat(60));
        console.log();
        console.log('📁 Generated Files:');
        if (fs.existsSync(templateExcel)) {
            console.log(`  1. ${templateExcel}`);
        }
        if (fs.existsSync(completedExcel)) {
            console.log(`  2. ${completedExcel}`);
        }
        console.log();
        console.log('📝 Next Steps:');
        console.log('  1. Open the Excel files to review the structure');
        console.log('  2. Template file can be shared with Analysts');
        console.log('  3. Completed file shows example submission with data');
        console.log();
    } catch (error) {
        console.error('❌ Error:', error.message);
        process.exit(1);
    }
}

main();
